src/libraries/System.Net.Security/tests/FunctionalTests/SslStreamNetworkStreamTest.cs

```diff
@@ -171,6 +171,239 @@ public async Task SslStream_NetworkStream_Renegotiation_Succeeds(bool useSync)
+                    Assert.NotNull(server.RemoteCertificate);
+                }
+                else
+                {
+                    Assert.Null(server.RemoteCertificate);
+                }
+                // Finish the client's read
+                await server.WriteAsync(TestHelper.s_ping);
+                await t;
+                // verify that the session is usable with or without client's certificate
+                await TestHelper.PingPong(client, server);
+                await TestHelper.PingPong(server, client);
+            }
+        }
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.SupportsTls13))]
+        [InlineData(true)]
+        [InlineData(false)]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public async Task SslStream_NegotiateClientCertificateAsyncTls13_Succeeds(bool sendClientCertificate)
+        {
+            bool negotiateClientCertificateCalled = false;
+            (SslStream client, SslStream server) = TestHelper.GetConnectedSslStreams();
+            using (client)
+            using (server)
+            using (X509Certificate2 serverCertificate = Configuration.Certificates.GetServerCertificate())
+            using (X509Certificate2 clientCertificate = Configuration.Certificates.GetClientCertificate())
+            {
+                SslClientAuthenticationOptions clientOptions = new SslClientAuthenticationOptions()
+                {
+                    TargetHost = Guid.NewGuid().ToString("N"),
+                    EnabledSslProtocols = SslProtocols.Tls13,
+                };
+                clientOptions.RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) => true;
+                clientOptions.LocalCertificateSelectionCallback = (sender, targetHost, localCertificates, remoteCertificate, acceptableIssuers) =>
+                {
+                    return sendClientCertificate ? clientCertificate : null;
+                };
+
+                SslServerAuthenticationOptions serverOptions = new SslServerAuthenticationOptions() { ServerCertificate = serverCertificate };
+                serverOptions.RemoteCertificateValidationCallback = (sender, certificate, chain, sslPolicyErrors) =>
+                {
+                    if (negotiateClientCertificateCalled && sendClientCertificate)
+                    {
+                        Assert.Equal(clientCertificate.GetCertHash(), certificate?.GetCertHash());
+                    }
+                    else
+                    {
+                        Assert.Null(certificate);
+                    }
+
+                    return true;
+                };
+
+                await TestConfiguration.WhenAllOrAnyFailedWithTimeout(
+                                client.AuthenticateAsClientAsync(clientOptions),
+                                server.AuthenticateAsServerAsync(serverOptions));
+                // need this to complete TLS 1.3 handshake
```

Comment:
Can you explain why this is needed for 1.3? Not obvious to me...

Reply:
When `AuthenticateAs*` is finished with TLS 1.3, the handshake is really not done. (unfortunately)
This is clearly visible on Windows when on first message, the "renegotiation" path is triggered and underlying TLS state will change.  So the next line will make sure that we are in fully operational and completed state. 

Now, when I try to trigger renegotiation e.g. post-handshake-authentication in this state it fails. 
In practice this should not matter as the server will receive request and that will get this completed AFAIK.

I was wondering if I could/should parametrize the test above but I decided not to as the PHA is really different beast as separate test method may be easier to update and understand. For Linux I only have prototype with TLS 1.2 so I don't know if TLS 1.3 will behave same way or not. It is certainly different set of APIs in OpenSSL.

Reply:
Just to make sure I understand this...

With TLS 1.3, exactly what TLS state can change after AuthenticateAs*? Is it that after AuthenticateAsServer, we don't necessarily have the client cert yet?

Or are there other cases too?

I'm wondering how a user is supposed to deal with this...

Reply:
It is the underlying  Schannel. it would essentially trigger the renegotiation path and calling write or another renegotiation would fail at that point. We would have certificate but the Schannel is still probably updating some crypto parameters. 

Normally this would not be visible to the user. we had to rework locking quite a bit in 5.0. So I padded many tests that do care about particular behavior with the ping/pong to make sure everything is in steady state. 
For TLS 1.2 and bellow we may for example receive SessionTickets. It does not matter too much in this case but it is another example when there is come extra processing after the handshake is done. 

I don't know how Linux will behave at this point for TLS 1.3 since it is using different mechanism. 
One option may be to remove this test and table TLS 1.3 as separate effort.

Reply:
Okay, I think I understand... let me make sure...

The issue here is that with TLS 1.3, the *initial* handshake isn't actually guaranteed to be done on the server when AuthenticateAsServer completes. (Aside: is this only on the server, or on the client too?)

In practice this is not visible to the user *except* for this case, right? E.g. if the client sends a cert in the initial handshake, it will be available as soon as AuthenticateAsServer completes, right? Are there any user-visible parameters that might change after AuthenticateAsServer completes?

In this case though, I think the issue is that we can't start the "renegotiation" (or really the TLS 1.3 PHA) until the initial handshake is fully completed, is that correct?

What happens if the user calls NegotiateClientCertAsync in this case, will it fail?

I wonder if we should just wait until the initial handshake completes. Is that possible? Is it hard?

Reply:
Yes, it was failing for me if you try renegotiation while in this weird state. I decided (at least for now) that this should not be problem in practice as the server will need receive and consume the request first so all this would actually finish. 
The detection may be tricky as it works different way on Linux. The ReplyOnReAuthenticationAsync is never called but I suspect the underlying OpenSSL state is similar as we saw some strange errors before reworking the locking.  I can dig through the TLS 1.3 RFC but I think some of the crypto piggybacks on first data to to cut down on roundtrips.

Reply:
> I decided (at least for now) that this should not be problem in practice as the server will need receive and consume the request first so all this would actually finish.

Yeah, I guess that's reasonable. It seems like the client is always going to send something once the handshake completes, anyway.

Should we add a TLS-1.3 specific test to ensure we fail properly if they call NegotiateClientCertAsync in this weird state?

Reply:
> I can dig through the TLS 1.3 RFC but I think some of the crypto piggybacks on first data to to cut down on roundtrips.

I think it would be really good to understand this better. I feel like users could get very confused here.

Reply:
agreed. I can ping the Schannel people as well.